driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -289,7 +291,13 @@ public void refreshSchema(String keyspace, String table) throws InterruptedExcep
             // At startup, when we add the initial nodes, this will be null, which is ok
             if (c == null)
                 return;
-            refreshSchema(c, keyspace, table, cluster, false);
+            refreshSchema(c, keyspace, table, cluster);
+            // If the table is null, we either rebuild all from scratch or have an updated keyspace. In both cases, rebuild the token map
+            // since some replication on some keyspace may have changed
+            if (table == null) {
+                cluster.submitNodeListRefresh();
+                cluster.nodeListRefreshRequestDebouncer.deliverEvents().get();
```

Comment:
You are absolutely right, we probably won't be able to block here waiting on this future. Conceptually, I think it is acceptable; but the metadata state will not always be accurate and we could have some tests failing more or less deterministically.